src/query/api/v1/handler/prometheus/native/complete_tags_test.go

```diff
@@ -202,3 +204,40 @@ func TestMultiCompleteTags(t *testing.T) {
 	actual := w.Header().Get(headers.LimitHeader)
 	assert.Equal(t, "max_fetch_series_limit_applied,abc_def", actual)
 }
+
+//nolint:dupl
+func TestCompleteTagsKillOnTimeout(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	// Setup storage
+	store := storage.NewMockStorage(ctrl)
+	store.EXPECT().CompleteTags(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(ctx context.Context,
+			query *storage.CompleteTagsQuery,
+			options *storage.FetchOptions,
+		) (*consolidators.CompleteTagsResult, error) {
+			<-ctx.Done()
+			// Prove the fact that we've passed along the timeout and that it's exceeded
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				return nil, ctx.Err()
+			}
+			return nil, nil
+		},
+	)
+
+	fb, err := handleroptions.NewFetchOptionsBuilder(
+		handleroptions.FetchOptionsBuilderOptions{Timeout: 1 * time.Millisecond})
```

Comment:
a less flaky way of doing this would be to explicitly cancel the ctx before you issue the request

Reply:
gonna follow up since I've got a completed green build :-)

Reply:
hmm so played around with this; because we wrap the context in the handler with a timeout, it doesn't respect an already. cancelled context by that point.

```
	cancelledCtx, cancel := context.WithCancel(context.Background())
	cancel()
	wrappedCtx, _ := context.WithTimeout(cancelledCtx, 1*time.Second)
	assert.NotNil(t, cancelledCtx.Err())
	assert.Nil(t, wrappedCtx.Err())
```

Reply:
weird. that's not what I see https://play.golang.org/p/8a_4SWOYKLQ

Reply:
hmmm yeah... will look more into this

Reply:
ah yeah, I think I was checking for the wrong error (deadline exceeded rather than cancelled). However I'm realizing that's a different test case all together (making sure the request context is used, which is separate from making sure the configured timeout is set).

I'll create those tests separately.

Reply:
https://github.com/m3db/m3/pull/3215